### PR TITLE
Change implementation of Bravyi-Kitaev

### LINKED ISSUE #259: Which Bravyi-Kitaev transform?
As discussed at #232 , our current implementation of the Bravyi-Kitaev transform comes from arXiv:1701.07072, and this definition is actually different from the one given by Bravyi and Kitaev in arXiv:quant-ph/0003137, which is the same as the one in arXiv:1208.5986. Apparently this fact had gone unnoticed until that discussion so I thought we should discuss now whether or not we should change our implementation. If we want to implement both, then I'm asking which one should be our "canonical" one and be assigned the function name `bravyi_kitaev`.

The two transforms actually correspond to different variants of the Fenwick tree data structure; the one in arXiv:quant-ph/0003137 is also called a binary indexed tree and it is the one that is actually used in practice because it can be implemented efficiently as an implicit data structure (elements are stored in an array and parent-child relationships computed using bitwise operations on the indices of the array).

One reason to change our implementation is that the one from quant-ph/0003137 can be performed faster. That's because the parity and update sets can be computed from the mode index using arithmetic (actually, bitwise) operations, rather than actually traversing a tree data structure.

Another nice property about the binary indexed tree version is that the form of the transformed operators depends less on the number of modes. For instance, the annihilation operator a_5 looks the same whether the total number of modes is, 9, 10, ..., 15, and then at 16 (the next power of 2) just one more X operator gets attached. On the other hand, with the current implementation, a_5 looks different for all these different total number of modes.

### PR DESCRIPTION
**Pull request by kevinsung**
Changes the implementation of the Bravyi-Kitaev transform from the one described in arXiv:1701.07072 to the original one described in arXiv:quant-ph/0003137, which is the same as the one described in arXiv:1208.5986.

Keeps the old implementation under the name `bravyi_kitaev_tree`.

Resolves #259 .

This change is good for four reasons:
1. The new implementation results in QubitOperators with (in many cases strictly) smaller many-body order (this was verified numerically for all systems up to size 1000 but not proven).
2. The new implementation is faster (uses bitwise operations to compute parity and update sets instead of traversing a literal tree data structure).
3. The new implementation has simpler and shorter code (no need for a separate file to store a classical data structure).
4. In the new implementation, the indices that a given annihilation operator acts on is less dependent on the total number of modes.

I realize this might take some time to review and I'm happy to answer any questions. I plan to eventually write something up explaining the subtleties here.

### CODE REVIEW

**Comment by babbush** [comment]
I don't think we should delete the Fenwick tree implementation. I think we should keep both. Perhaps we can call the new code Bravyi_Kitaev and call the old code "Fenwick_Tree" or something like that.

**Comment by kevinsung** [comment]
What do you think of having the "Fenwick tree" implementation available only through the binary code transforms functionality?

**Comment by babbush** [comment]
We already have nice code for it. I suggest we keep what we have. Some might find it useful.

**Comment by kevinsung** [comment]
Ok I propose the name `bravyi_kitaev_tree` in recognition of the literal tree data structure it uses and that Fenwick never actually proposed to implement his data structure that way (his original data structure uses bitwise operations like in the original Bravyi-Kitaev and my implementation here).

**Comment by babbush** [comment]
that's fine

**Comment by kevinsung** [comment]
Ok I've done that. I'll do think this is ugly though because the test files for the two transforms are identical, and in fact much of the code in the two implementations is identical too. I think the ideal way to make the old version available is only as a BinaryCode (but I won't do that now).

**Comment by kevinsung** [comment]
Something's definitely wrong with coveralls... now it's penalizing me for increasing coverage!

File: src/openfermion/transforms/_bravyi_kitaev.py [hunk 1]
```diff
@@ -11,15 +11,13 @@
...
 """Bravyi-Kitaev transform on fermionic operators."""
-from __future__ import absolute_import
 
 from openfermion.ops import QubitOperator
-from openfermion.transforms._fenwick_tree import FenwickTree
+from openfermion.utils import count_qubits
 
 
 def bravyi_kitaev(operator, n_qubits=None):
     """Apply the Bravyi-Kitaev transform and return qubit operator.
```
Comment (babbush) [inline src/openfermion/transforms/_bravyi_kitaev.py]: Especially given all the confusion we've had, please give more description of which BK transform and ideally provide references to both Seeley and Bravyi+Kitaev

**Comment by babbush** [comment]
Are you actually increasing test coverage though? It seems like the error might be in the "+" sign?

File: src/openfermion/transforms/_bravyi_kitaev.py (hunk 1 above)
Comment (kevinsung) [reply src/openfermion/transforms/_bravyi_kitaev.py]: Done.

**Pull request merged**